File: 프로그래머스/lv2/42586. 기능개발/기능개발.js
function solution(progresses, speeds) {
    let answer = [];
    let cnt = 0;
    while(progresses.length>0){
        for(let i=0; i < progresses.length; i++){
            progresses[i] += speeds[i];
        }
        cnt=0;
        while(progresses[0] >= 100){
            cnt++;
            progresses.shift();
            speeds.shift();
        }
        if(cnt > 0) answer.push(cnt);
    }
    
    return answer;
}

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        List<Integer> answer = new ArrayList<>();
        Deque<Integer> deque = new ArrayDeque<>();
        
        for(int i = 0; i < progresses.length; i++){
            int progress = progresses[i];
            int count = 0;
            
            while(progress < 100){
                count++;
                progress += speeds[i];
            }
            
            deque.add(count);
        }
        
        
        int count = 0;
        int number = 0;
        while(!deque.isEmpty()){
            if (number >= deque.peek()) {
                deque.poll();
            } else {
                number = deque.poll();
            }

            count++;
            
            if(deque.isEmpty()){
                answer.add(count);
            } else if(deque.peek() > number){
                answer.add(count);
                count = 0;
                number = 0;
            }
        }
        
        return answer.stream().mapToInt(i -> i).toArray();
    }
}